ocean_provider/validation/RBAC.py

```diff
@@ -0,0 +1,170 @@
+
+    @property
+    def assets(self):
+        return self._assets
+
+    @property
+    def request(self):
+        return self._request
+
+    @property
+    def algorithms(self):
+        return self._algorithms
+
+    @property
+    def additional_inputs(self):
+        return self._payload["additionalInputs"]
+
+    def fails(self):
+        return False
+
+    def get_compute_dict(self):
+        return {
+            "dids": [
+                {
+                    "did": asset.did,
+                    "serviceId": ServiceTypesIndices.DEFAULT_COMPUTING_INDEX,
+                }
+                for asset in self.assets
+            ],
+            "algos": [
+                {
+                    "did": algorithm.did,
+                    "serviceId": ServiceTypesIndices.DEFAULT_COMPUTING_INDEX,
+                }
+                for algorithm in self.algorithms
+            ],
+            "additionalDids": [
+                {
+                    "did": additional_input["documentId"],
+                    "serviceId": additional_input["serviceId"],
+                }
+                for additional_input in self.additional_inputs
+            ],
+        }
+
+    def get_dids(self):
+        return {
+            "dids": [
+                {
+                    "did": asset.did,
+                    "serviceId": ServiceTypesIndices.DEFAULT_ACCESS_INDEX,
+                }
+                for asset in self.assets
+            ],
+        }
+
+    def build_payload(self):
+        payload = {
+            "eventType": self.action,
+            "component": self.component,
+            "providerAddress": self.provider_address,
+            "credentials": self.credentials,
+        }
+        # builds actions like build_encrtyptUrl_payload to update the dictionary
+        # with request - specific key-values.
+        payload.update(getattr(self, f"build_{self.action}_payload")())
+        return payload
+
+    def build_encryptUrl_payload(self):
+        message = "encryptUrl" + json.dumps(self.credentials)
+        signature = sign_hash(msg_hash(message), get_provider_wallet())
+
+        return {"signature": signature}
+
+    def build_initialize_payload(self):
+        dids = self.get_dids()["dids"]
+        message = "initialize" + json.dumps(self.credentials)
+        signature = sign_hash(msg_hash(message), get_provider_wallet())
+        return {
+            "signature": signature,
+            "dids": dids,
+        }
+
+    def build_access_payload(self):
+        dids = self.get_dids()["dids"]
+        message = "access" + json.dumps(self.credentials)
+        signature = sign_hash(msg_hash(message), get_provider_wallet())
+        return {
+            "signature": signature,
+            "dids": dids,
+        }
+
+    def build_compute_payload(self):
+        dids = self.get_compute_dict()["dids"]
+        algos = self.get_compute_dict()["algos"]
+        additional_dids = self.get_compute_dict()["additionalDids"]
```

Comment:
why is get_compute_dict a different function?
- it is only used here
- and it is used piecemeal...

move the functionality here, remove the function.
dids will be self.get_dids either way
you just need to add algos and additional dids.

Reply:
I understand the dids part is different, ok 👍 .
Still, please move the functionality from compute_dict here and do not build it as a dict if you don't use it as a dict. If the only difference between the return dict and the compute dict that you build is one key, why do you not add it with name_of_dict["extra_key"] = variable ?

Reply:
this is still not handled.